### DISCUSSION TITLE
using devices which are grouped / racked and unfolding them

### DISCUSSION THREAD

**Discussion by markusschloesser** [markusschloesser/MackieC4_P3#8] [Ideas]
# code snippet from "The Rabbits" at Ableton forum for RACKS
`
def get_device_list(self, container):

    # add each device in order. if device is a rack, process each chain recursively
    # don't add racks that are not showing devices.
    lst = []
    for dev in container:
        lst.append(dev)
        if dev.can_have_chains:  # is a rack and it's open
            if dev.view.is_showing_chain_devices:
                for ch in dev.chains:
                    lst += self.get_device_list(ch.devices)
    return lst
`

# Called this way.
`device_list = self.get_device_list(current_track.devices)`

now have to make a button, probably best to use v2.control_surface.control.button and then need to map this to a free button in channel strip mode. Alternatively use selection of device for unfolding/folding

**Comment by BilldarBagdar** [comment]
I need to refamiliarize myself with the code, and your updates.  But this seems to be something the existing "database" of nested lists wouldn't necessarily be able to handle.  Essentially that database is hard coded to 128 items in each list.  127 (normal + return) + master track = 128 tracks, 128 devices per track, and 128 parameters per device.  When "track groups" or "device chains" open and close, the track list and device lists can change dramatically, and possibly cross that "array index boundary" at 129.

With that said, I think IIRC, that the "encoder button press" event is not really implemented in the "device view", so I think it would be pretty cool if when you were in device view and "next device" landed you on a "device chain" device, you could press a button to expand that chain and land on another "layout" that represents a "chain view".  Like the way you can see the first few devices on a track in channel view, and jump to device view of that device by clicking the encoder button; you could see the first few devices in the expanded chain in device view and jump to device view of that device by similarly clicking the encoder button..

I don't know about you, but I have never had more than 4 or 5 return tracks in a Live session.  I don't understand the need for supporting 12 "send levels" in Channel mode.  Perhaps a few of those "" Send Level" encoders could be repurposed.

**Reply by markusschloesser** [reply]
> this seems to be something the existing "database" of nested lists wouldn't necessarily be able to handle. Essentially that database is hard coded to 128 items in each list. 127 (normal + return) + master track = 128 tracks, 128 devices per track, and 128 parameters per device. When "track groups" or "device chains" open and close, the track list and device lists can change dramatically, and possibly cross that "array index boundary" at 129.

My goal is to get everything from the LOM, so that I don't have the static "db". By getting stuff from the LOM, updates of group tracks, nested devices, chains etc. should come automatically .

> With that said, I think IIRC, that the "encoder button press" event is not really implemented in the "device view",

I'm not sure I follow?

>  I think it would be pretty cool if when you were in device view and "next device" landed you on a "device chain" device, you could press a button to expand that chain and land on another "layout" that represents a "chain view".

That's exactly what I want to do! :D

> I don't know about you, but I have never had more than 4 or 5 return tracks in a Live session. I don't understand the need for supporting 12 "send levels" in Channel mode. Perhaps a few of those "" Send Level" encoders could be repurposed.

IMHO the amount of "sends" also should come from the LOM and not be set statically.

**Reply by BilldarBagdar** [reply]
There are always 32 encoders and "you might as well show something".  To me having slots available by default in channel view is not a bad thing per se.  I mean "reserving" a whole row of 8 encoders for the first 8 devices on any track isn't a bad thing (as long as the script can handle "encoder activity" associated with non-existent devices).  But a whole row + 4 =  12 encoders dedicated to "Sends" seems excessive to me. I was thinking only the + 4 part would make the bottom row 4 "sends" + "Mute", "Arm", "Pan", and "Volume".  which would free up the 8 encoders in the third row for some other channel controls instead of "sends", but what kind of controls would we map to those 8 third row encoders?

**Comment by markusschloesser** [comment]
see #16

**Comment by markusschloesser** [comment]
> There are always 32 encoders and "you might as well show something".

I disagree, showing only something when there actually is something to show, reduces complexity (right now the showing of "none" instead of an empty display is imho a bad approach). I'd rather have certain encoders unused and instead have a logical grouping of functionality into different pages/banks of parameters (maybe in the future with "Page/Bank Names"). Don't know if that's feasible though in terms of what we get from the LOM. If there's such a "logical grouping" it might be in the "Push" scripts.

>  I mean "reserving" a whole row of 8 encoders for the first 8 devices on any track isn't a bad thing (as long as the script can handle "encoder activity" associated with non-existent devices). 

reserving the space makes sense, but the current approach of showing non-existent sends doesn't. Those should be empty and unlit. Ultimately we could introduce paging to the "sends" row by for example using vpot encoder press, when there are more then 7 sends (and use the eighth encoder for paging.

>  but what kind of controls would we map to those 8 third row encoders?

as that concerns the channel view, I suggest 

1. "fold/unfold" for group/nested tracks. 
2. maybe routing? 
3. midi inputs for midi channels

**Reply by BilldarBagdar** [reply]
I agree - essentially, I suspect the only reason you see "None" is because that's how I fixed the problems I was having with some "vpot_parameter" values that were getting accessed in the original `on_display_update_timer()` method last year.  IIRC, before the script showed "None", it was filling the log with vomit.

> You might as well show something

I also agree with this.  "Showing nothing" when there is nothing to show is perfectly legitimate program behavior.  What I meant was just that the encoders are physically present and they might as well be mapped.  What I wouldn't want to see though would be where some encoders could do "double duty" within a "mode" of the script.  What I mean is, for example, one track has 12 sends and another track has 2 sends, and the track with 2 sends has "extra functions" mapped to some or all of the other 8 encoders used by the first Track's 12 sends.  We should support, for example (the first) 2, 4, 8, 12 sends (on a Track) and stick with that plan across the board.

**Reply by markusschloesser** [reply]
couple of things:
1. in chan strip where the devices are listed (and selectable), it would be cool to a) show device name on top row (easy, done) and the device status (on/off) on bottom row. Plus of course the encoder rings enlighted or not depending on status. As that's PER DEVICE, do you have any idea how to get that status easily without building new indexes?
2. have a look at https://github.com/markusschloesser/AbletonRemoteScriptsPY/blob/main/Push2/track_selection.py, https://github.com/markusschloesser/AbletonRemoteScriptsPY/blob/main/ableton/v2/control_surface/device_provider.py  and https://github.com/markusschloesser/AbletonRemoteScriptsPY/blob/main/ableton/v2/control_surface/components/device_navigation.py I think we can learn and probably reuse a LOT from those 3 w re to db, chains, listeners etc. For example I have been trying to figure out if and how to use `FlattenedDeviceChain` for the "DB" and then `device_to_appoint(device)` for selection. But cannot figure it out atm.
3. I looked at the touchable scripts yesterday and even I (!) can tell that they are very messy and impossible to maintain. They are at least 10 the size of ours. Reason: they do and define everything by hand. So taking inspiration from Hanz Petrov's articles (see wiki here), I think the way forward is definitely by using the v2 framework (even though right now the whole **kwargs shit is messing with my still limited understanding 🤪 but I will get there :-)

**Comment by markusschloesser** [comment]
done for Tracks (https://github.com/markusschloesser/MackieC4_P3/issues/46)

**Comment by markusschloesser** [comment]
closing/locking this, as it is an issue not a discussion
